### ISSUE TITLE
Internet browser for JDownloader

### ISSUE THREAD

**Issue by clx81** [jlesage/docker-jdownloader-2#3]
Hello,
To solve some captcha, JD needs access to a web browser, ie. Turbobit.
It shoud'be nice to integrate a light browser inside the container.
I will try to do it.
Many thanks for your nice work.
Paul

**Comment by jlesage** [comment]
Thanks for the suggestion.  I will look at this!

**Comment by rix1337** [comment]
Just for debugging: For most captchas the integrated phantomjs should do the work. 

If I remember correctly, a real browser is only needed for exotic captchas, or if something goes wrong. Would be a good fallback, though!

**Comment by jlesage** [comment]
Do you have any example of a download link that would require a real browser?

**Comment by rix1337** [comment]
Not anymore, sadly. @clx81 since you brought it up: any example?

**Comment by jlesage** [comment]
For reference, this link has more information about captchas that need browser: https://support.jdownloader.org/Knowledgebase/Article/View/42/14/jd-opens-my-browser-to-display-captchas

**Comment by clx81** [comment]
Hello Jocelyn,
This link need browser : http://turbobit.net/ozmzsk9hxi2c.html

**Comment by rix1337** [comment]
To add further, some settings might need tweaking:

In the advanced settings:

`GeneralSettings: Browser Command Line
CommandLine to open a link in a browser. Use %s as wildcard for the url`

is currently set to `["/usr/bin/firefox","%s"]` on my setup.

**Comment by jlesage** [comment]
In this scenario, a full web browser is not required...  Ideally, I would like to find and use a light-weight browser that support the captchas.  A browser like firefox or chromium would double (if not more) the size of the container image!

**Comment by clx81** [comment]
There is Dillo running on Alpine, but I don't know if it can do the job !

**Comment by jlesage** [comment]
Tried, but doesn't properly display captchas (at least the SweetCaptcha one).

**Comment by clx81** [comment]
I agree that Firefox will increase the container image but if you run JD2 for a few days with some downloads the memory size increased up to 900 Mb. We see that also when running it out of docker (verified on Debian 8)

**Comment by jlesage** [comment]
You talk about the RAM or disk space?

**Comment by clx81** [comment]
I talk about the RAM, I think that JD2 don't flush the RAM until reboot.

**Comment by mapleshadow** [comment]
Please support built-in browser, otherwise it can not pass captcha.
At present, you can use the Android client, the normal input captcha.
But that's not the perfect way.

**Comment by rix1337** [comment]
My setup is highly automated and can even solve recaptcha v2 with the integrated shadowjs. Am using 9kw.eu for this.
This is working fine as-is. No browser required...

**Comment by mapleshadow** [comment]
@rix1337 Thank you very much.

**Comment by rix1337** [comment]
@jlesage Thank you for the frequent updates!

**Comment by grundscm** [comment]
Hello, it is possible to implement a firefox inside container? I tried it  without any success.
Thanks a lot

**Comment by jlesage** [comment]
You can install it by executing `add-pkg firefox-esr` in the container.  But like others said a web browser should not be required.

**Comment by jlesage** [comment]
I remember having similar issue when I evaluated the possibility of integrating a browser.  I think there is some incompatibility between JDownloader, which needs glibc, and Firefox, which is compiled with musl.

**Comment by rix1337** [comment]
another info: if you dont want to user solver services (9kw, etc.) you can enable "solve by app/myjdownloader".

This works for all captcha types and does not require a browser aswell. 

You just need a phone or desktop browser at hand when a captcha appears.

**Comment by grundscm** [comment]
@rix1337 Thanks for the additional info!
Maybe there is another browser easier to implement in the container?
I think "Copy/Paste"(mobile phone us) would be easier if i have both in one container.

**Comment by rix1337** [comment]
none that I know of.. maybe try the captcha through myjdownloader option for now.

**Comment by diyfr** [comment]
You have [PhantomJs](http://phantomjs.org/) 
Question: can not directly paste a link, we must go through the clipboard button at the top right ?

**Comment by jlesage** [comment]
Sorry what do you mean exactly?  Firefox is not part of the container image...

**Comment by Draic** [comment]
oh my fault, sorry. This is the wrong docker :D. mixed two nearly identical ones. ignore me

**Comment by SubrataM3** [comment]
Hi @jlesage 
I have tried installing firefox-esr package on the image but its working strangely.
If i log into the shell and run firefox command i get firefox window on my browser and the browsing works without any issue.
But if try to open firefox clicking any link on jdownloader different kind of error arises.

Example error 1:
`[Parent 6507] WARNING: pipe error (38): Connection reset by peer: file /home/buildozer/aports/community/firefox-esr/src/firefox-52.9.0esr/ipc/chromium/src/chrome/common/ipc_channel_posix.cc, line 322`

Example error 2:
```
Gdk-Message: 18:44:46.182: firefox: Fatal IO error 11 (Resource temporarily unavailable) on X server :0.

[Child 5799] ###!!! ABORT: Aborting on channel error.: file /home/buildozer/aports/community/firefox-esr/src/firefox-52.9.0esr/ip.cpp, line 2152
[Child 5799] ###!!! ABORT: Aborting on channel error.: file /home/buildozer/aports/community/firefox-esr/src/firefox-52.9.0esr/ip.cpp, line 2152
```

I am not sure whats going on here.
I think it will be really helpful if you can add a seperate image with firefox for those who don't care about the image size.

Thanks.

**Comment by jlesage** [comment]
Try to create the container with the following parameter: `--shm-size 2g`
This is usually required to properly runs firefox.

**Comment by calvinbui** [comment]
i had issues with Firefox, Dillo and Midori. Chromium worked for me. In Advanced Settings i put:

```
GeneralSettings: Browser Command Line -> ["/usr/bin/chromium-browser", "--no-sandbox", "%s"]
BrowserCaptchaSolver: Browser Commandline -> ["/usr/bin/chromium-browser", "--no-sandbox", "%s"]
```

**Comment by drod3763** [comment]
I am running into a similar issue regarding browsers when trying to use my real-debrid account. It creates a popup for Oauth which won't open anything for the obvious reasons. Is there a workaround for this?

![Screen Shot 2020-05-06 at 11 18 11 AM](https://user-images.githubusercontent.com/7785716/81195090-6242ea00-8f8b-11ea-8d88-12940d95441f.png)

**Comment by trulow** [comment]
Any updates or workarounds to this issue? I can't get realdebrid to authenticate because I keep getting an oauth error

**Comment by PavelD** [comment]
> Any updates or workarounds to this issue? I can't get realdebrid to authenticate because I keep getting an oauth error

I spend some time to finalize it so here are my steps:

I created simple docker file like that:

```
FROM jlesage/jdownloader-2

RUN apk add chromium
```

That's all for the installation itself.

For configuration is had to tweak a few thinks.

I had to define the `container_name` in docker compose file to avoid report of locked profile.

Then in the `Advanced Settings i changed configuration according to @calvinbui post. (not sure what is doing the --no-sandbox, it's reported by current versions of chromium, but it's still working).

As well i install the extension to chromium after the fist run. The chromium profile stays is the mounted folder so it is necessary to do it only once.

After this change the size of the docker image has changed from 224MB to 501MB but solving captcha in browser is working.

Hope this brief description will help.

**Comment by trulow** [comment]
For anyone who is also looking to add real debrid to their installation but can't due to the issues with oauth. I was able to solve it by using this method provided by jdownloader

How to copy your account settings into a headless JD installation:
    1. Add all of your desired accounts to a JD installation with GUI (alternatively, copy the account database from your headless installation into your GUI installation, see steps below).
    2. Define your desired account settings.
    3. Close your headless JDownloader.
    4. Copy your account database from [Your GUI-JD install dir]/cfg/org.jdownloader.settings.AccountSettings.accounts.ejs to your headless installation.
        ⁃ Click on Help>About>Installation Directory to find your install directory
Your Account settings should now be applied to your headless installation.

https://support.jdownloader.org/Knowledgebase/Article/View/headless-settings

In my case, I copied the org.jdownloader.settings.AccountSettings.accounts.ejs file to the config/cfg which was defined in the docker run command

**Comment by jlesage** [comment]
Accounts can also be added via MyJDownloader (https://my.jdownloader.org).  So this is probably the easiest way to add a read-debrid account.
